CFP: Pod MAC Address Specification

chaunceyjiang opened the pull request:
Specify the MAC address of the pod in the annotation, store the MAC address in the endpoint, and then cilium-cni obtains the specified MAC address by querying the endpoint and sets the MAC address of the interface inside the pod.



Fixes: #22119

```release-note
Add support for the `cni.cilium.io/mac-address` annotation on Pod resources to control the L2 address used for Pod communication.
```

joestringer commented on daemon/cmd/endpoint.go:
```diff
@@ -561,7 +577,7 @@ func putEndpointIDHandler(d *Daemon, params PutEndpointIDParams) (resp middlewar
 	ep.Logger(daemonSubsys).Info("Successful endpoint creation")
 
-	return NewPutEndpointIDCreated()
+	return NewPutEndpointIDOK().WithPayload(fmt.Sprintf("%d", ep.ID))
```
I don't understand why we would need to change the API response code in order to implement this feature. `200` is also less accurate than the current `201` we are using, which typically indicates not only that the request was accepted, but that resources were created based on the request.

joestringer replied:
Presumably the existing return code could just be used with the new payload?

Also, for forwards compatibility, the result should probably be a structure rather than a string.

chaunceyjiang replied:
> I don't understand why we would need to change the API response code in order to implement this feature.

Because I stored the MAC address in `endpoint`, I need to query `endpoint` to know the MAC address. However, the method `EndpointGet` for querying `endpoint` requires knowing the `endpointID`, and currently there is no way to know the `endpointID`. Therefore, I modified the `EndpointCreate` method to return the ID of newly created `Endpoint`.

chaunceyjiang replied:
> Presumably the existing return code could just be used with the new payload?

I didn't quite understand what you said, could you describe it more clearly? Thank you very much.

joestringer commented on plugins/cilium-cni/chaining/generic-veth/generic-veth.go:
```diff
@@ -216,7 +216,7 @@ func (f *GenericVethChainer) Add(ctx context.Context, pluginCtx chainingapi.Plug
 	})
 
-	err = cli.EndpointCreate(ep)
+	_, err = cli.EndpointCreate(ep)
```
Shouldn't Cilium always control the MAC address of the devices if this feature is enabled? I think it would be confusing if this doesn't work in generic-veth mode as well.

joestringer commented on plugins/cilium-cni/main.go:
```diff
@@ -590,12 +592,25 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
+		if err != nil {
+			return fmt.Errorf("unable to set MAC address on interface %s: %w", args.IfName, err)
+		}
+	}
```
I think that this would be more compatible if the plugins do `if epID != nil { ... }` instead. This way, if the plugin version is mismatched with the agent version, the plugin will gracefully fall back to not implementing the MAC change.

Otherwise, I could see that if the user upgrades then downgrades Cilium, there could be a period where the agent becomes an older version, but the plugin is a new version, and every endpoint create would hit the `epID == nil` error case for some time.

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -146,6 +146,8 @@ paths:
         '201':
           description: Created
+          schema:
+            "$ref": "#/definitions/EndpointID"
```
Did you consider whether to reuse the existing `Endpoint` object?

It looks like there are no required fields for that object type, so you should be able to construct a response similar to your current solution with only the endpoint ID. I think that this would intuitively make sense to have the response to the endpoint creation be an endpoint object.

The benefit would be that you could drop the `EndpointID` object, which would keep the API simpler.

joestringer commented on daemon/cmd/endpoint.go:
```diff
@@ -413,6 +414,21 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
+			}
+			ep.SetMac(m)
+		}
+	}
```
I think that these changes should be combined with the `fetchK8sMetadataForEndpoint` code below. No need to do that step twice.

joestringer reviewed (changes requested):
Good progress :+1: , I have a few suggested changes below to simplify the implementation and improve upgrade compatibility.

joestringer commented on daemon/cmd/endpoint.go:
```diff
@@ -453,6 +454,15 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
 				}).Warningf("Endpoint has %s annotation, but BPF bandwidth manager is disabled. This annotation is ignored.",
 					bandwidth.EgressBandwidth)
 			}
+			if hwAddr, ok := annotations["cni.cilium.io/hwAddr"]; ok {
```
One more minor nit, should this be named something like `cni.cilium.io/macAddr`? I think `hw` expands to `hardware` which is not strictly true given we're talking about virtual device configuration.

chaunceyjiang replied:
I borrowed from [Calico](https://docs.tigera.io/calico/latest/networking/configuring/pod-mac-address), not sure if this is the best practice.

joestringer replied:
The label is already different, I'm not sure if there is much point in retaining a partial name match :shrug:

chaunceyjiang replied:
OK，I will change `cni.cilium.io/hwAddr` to `cni.cilium.io/macAddr` later.

joestringer commented on daemon/cmd/endpoint.go:
Overall I think there are two more things I'd like to see as part of this PR:

1. Documentation for the new feature
2. Tests to validate that the new feature works and does not regress.

As part of documentation, this should also describe any caveats. For instance, I believe that the MAC address must be specified in this annotation during Pod creation, and Cilium will *not* respect changes to the annotation after Pod creation.

chaunceyjiang replied:
> Documentation for the new feature

I haven't written any documentation in cilium yet, can you remind me where this feature should be written? Thank you very much.

joestringer replied:
There are some other network-related features documented in `Documentation/network`. The files are listed in `Documentation/index.rst` (without the `.rst` filename suffix). There is some guides around documentation here, including style and how to view it locally and iterate: https://docs.cilium.io/en/stable/contributing/docs/#docs-guide

joestringer commented on daemon/cmd/endpoint.go:
```diff
@@ -573,7 +583,7 @@ func putEndpointIDHandler(d *Daemon, params PutEndpointIDParams) (resp middlewar
 	ep.Logger(daemonSubsys).Info("Successful endpoint creation")
 
-	return NewPutEndpointIDCreated()
+	return NewPutEndpointIDCreated().WithPayload(ep.GetModel())
```
Heads up for API reviewers that this will increase the size of the endpoint creation response for every Pod creation / CNI ADD event, but I don't think this should be too costly.

joestringer reviewed (commented):
The general code approach looks good to me now, I'll invite also the other reviewers to take a look. Below I have a few more comments on the specifics around educating users about this feature and about how we ensure that the feature continues to work in future releases.

christarazi commented on test/k8s/pod_mac_address.go:
```diff
@@ -0,0 +1,70 @@
+			Expect(demoPods.Items).To(HaveLen(1))
+			By("Checking the pod's mac address")
+			kubectl.Exec(fmt.Sprintf("kubectl exec -it %s -c %s -- ip link show |grep \"%s\" | wc -l",
+				demoPods.Items[0].Name, demoPods.Items[0].Spec.Containers[0].Name, demoPods.Items[0].Annotations["cni.cilium.io/macAddr"])).
```
I think this can be simplified by using `kubectl.ExecPodCmd`.

chaunceyjiang replied:
Updated.

christarazi reviewed (approved):
ci-structure changes LGTM pending one minor nit.

chaunceyjiang committed:
Pod MAC Address Specification

Fixes: #22119

Specify the MAC address of the pod in the annotation, store the MAC address in the endpoint,
and then cilium-cni obtains the specified MAC address by querying the endpoint and sets the MAC address of the interface inside the pod.

Signed-off-by: chaunceyjiang <chaunceyjiang@gmail.com>

aanm commented on daemon/cmd/endpoint.go:
```diff
@@ -456,6 +457,15 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
 				}).Warningf("Endpoint has %s annotation, but BPF bandwidth manager is disabled. This annotation is ignored.",
 					bandwidth.EgressBandwidth)
 			}
+			if hwAddr, ok := annotations["cni.cilium.io/macAddr"]; ok {
```
IIRC the annotations are not camel case, I would suggest to change it to :
```
"cni.cilium.io/mac-address"
```

aanm replied:
Can we add the annotation in [here](https://github.com/cilium/cilium/blob/0f6f8980949276327fd77305d3ee2e05c268d573/pkg/annotation/k8s.go) and use a `const` instead?

aanm commented on .github/actions/ginkgo/main-focus.yaml:
```diff
@@ -212,6 +213,10 @@ include:
   - focus: "f20-kafka"
     cliFocus: "K8sKafkaPolicyTest"
```
`f20-kafka` is the test that is currently taking least amount of time to run. I would suggest to add the new test here.

```
cliFocus: "K8sKafkaPolicyTest|K8sSpecificMACAddressTests"
```

aanm commented on .github/actions/ginkgo/main-focus.yaml:
```diff
@@ -212,6 +213,10 @@ include:
+  ###
+  # K8sSpecificMACAddressTests Check whether the pod is created Checks the pod's mac address
+  - focus: "f21-specific-mac-address"
+    cliFocus: "K8sSpecificMACAddressTests Check whether the pod is created Checks the pod's mac address"
```
Let's drop this and add it to the f20-kafka

gandro commented on daemon/cmd/endpoint.go:
```diff
@@ -456,6 +457,15 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
+						WithError(err).Error("Unable to parse MAC address")
+					return invalidDataError(ep, err)
+				}
+				ep.SetMac(m)
```
With https://github.com/cilium/cilium/pull/26894/ we actually support more than one endpoint per pod. While lots of the logic to support multi-network is not yet upstreamed, it is clear that this is a use-case we are going to support.

Thus we need to think about the meaning of the pod annotation in case there are multiple endpoints per pod. 
Every endpoint will have its own interface, so we don't want to set the mac address the same for all interfaces. 

To be forward-compatible, I think we should check `DisableLegacyIdentifiers==false` before parsing the annotation and setting the MAC address here. By definition, only a single endpoint can have `DisableLegacyIdentifiers=false`, so that way we ensure that only one endpoint per pod has this custom MAC address set. This is also compatible with the current version of the CNI plugin which creates a single endpoint with `DisableLegacyIdentifiers=false`.

gandro replied:
Long-term, we also need to answer the question: How should the annotation look like if the pod has multiple interfaces? My above proposal basically suggests to only use the MAC for the "primary" endpoint, but a truly multi-network aware implementation might want to be able to define one MAC per interface.

chaunceyjiang replied:
> With https://github.com/cilium/cilium/pull/26894 we actually support more than one endpoint per pod. 

I haven't seen this feature before. Thank you for your review.


> To be forward-compatible, I think we should check DisableLegacyIdentifiers==false before parsing the annotation and setting the MAC address here.

Got it.

> How should the annotation look like if the pod has multiple interfaces? My above proposal basically suggests to only use the MAC for the "primary"


So, do we define annotation as this 

```
cni.cilium.io/mac-address: ["e2:9c:30:38:52:62", "e2:9c:30:38:52:63",]
```

or like this



``` 
cni.cilium.io/primary-mac-address: "e2:9c:30:38:52:61"  # I believe there should only be one "primary" endpoint.
cni.cilium.io/secondary-mac-address: ["e2:9c:30:38:52:62"] 
``` 


I feel the second option is more expandable, what do you think?

gandro replied:
Networks will probably be named, so I think we probably want to assign mac addresses based on the network name, e.g. `cni.cilium.io/mac-address: net1=e2:9c:30:38:52:62, net2=e2:9c:30:38:52:63` - but I think that's a problem for a later date.

For now I think it's fine if we start with a solution that just looks at `DisableLegacyIdentifiers` and only sets the mac if `DisableLegacyIdentifiers==false`. This will ensure it will only apply to the "primary" endpoint (whatever its name is).

chaunceyjiang replied:
>. For now I think it's fine if we start with a solution that just looks at DisableLegacyIdentifiers and only sets the mac if DisableLegacyIdentifiers==false. 

Done.

gandro reviewed (changes requested):
Thanks for the PR - I've left some feedback to be forward-compatible with the upcoming multi-network feature (c.f. also [KEP draft](https://docs.google.com/document/d/17LhyXsEgjNQ0NWtvqvtgJwVqdJWreizsgAZHWflgP-A/edit))

gandro reviewed (approved):
Awesome, thanks!

joestringer commented on Documentation/network/pod-mac-address.rst:
```diff
@@ -0,0 +1,69 @@
+Some applications bind software licenses to network interface MAC addresses.
+Typically, the solution is very complex and requires things like Docker within
+the cluster.
+
```
nit, I don't think that this sentence provides much clarity for the motivation of the feature. I think that the text would be just as clear without this.

```suggestion
```

joestringer commented on Documentation/network/pod-mac-address.rst:
```diff
@@ -0,0 +1,69 @@
+
+
+Specific MAC Address
+#####################
```
```suggestion
Configuring the address
#######################
```

joestringer commented on Documentation/network/pod-mac-address.rst:
```diff
@@ -0,0 +1,69 @@
+
+
+Annotate the pod with ``cni.cilium.io/mac-address`` set to the desired MAC address.
+After adding the annotation to the pod, it appears like so.
```
```suggestion
For example:
```

joestringer commented on Documentation/network/pod-mac-address.rst:
```diff
@@ -0,0 +1,69 @@
+      name: busybox
+      namespace: default
+
+After the pod runs successfully, we can query the pod's eth0 interface.
```
If the user needs to do something, we should instruct them to do it in simple terms. We shouldn't assume that they do something via wording like "after the pod runs successfully".

```suggestion
Deploy the Pod. Cilium will configure the mac address to the first interface in the Pod automatically.
```

chaunceyjiang replied:
Updated.

joestringer commented on Documentation/network/pod-mac-address.rst:
```diff
@@ -0,0 +1,69 @@
+
+.. warning::
+    The MAC address must be specified in this annotation during Pod creation,
+    and Cilium will not respect changes to the annotation after Pod creation.
```
I think that this is critical information, and should be worked into the text directly rather than adding a warning callout at the end of the document.

joestringer replied:
```suggestion
```

joestringer commented on Documentation/network/pod-mac-address.rst:
```diff
@@ -0,0 +1,69 @@
+the cluster.
+
+Therefore, cilium provides the ability to specific MAC addresses
+for pods instead of letting the operating system allocate them.
```
```suggestion
Cilium provides the ability to specific MAC addresses for pods
at deploy time instead of letting the operating system allocate them.
```

joestringer commented on Documentation/network/pod-mac-address.rst:
```diff
@@ -0,0 +1,69 @@
+The MAC address configured by the annotation described here will be visible
+from within the container on the eth0 interface. Since it is isolated to the
+container it will not collide with any other MAC addresses assigned to other
+pods on the same node.
```
```suggestion
Cilium will configure the MAC address for the primary interface inside a
Pod if you specify the MAC address in the ``cni.cilium.io/mac-address`` before
deploying the Pod. This MAC address is isolated to the container so it will
not collide with any other MAC addresses assigned to other Pods on the same
node. The MAC address must be specified *before* deploying the Pod.
```

joestringer replied:
Minor nit here: The naming of `eth0` for the primary interface is distribution-specific for the container image, so we should prefer `primary interface` and only use `eth0` as an example.

joestringer commented on Documentation/network/pod-mac-address.rst:
```diff
@@ -0,0 +1,69 @@
+
+
+
+
```
nit: extraneous whitespace
```suggestion
```

joestringer commented on Documentation/network/pod-mac-address.rst:
```diff
@@ -0,0 +1,69 @@
+Use a Specific MAC Address for a Pod
+************************************
+
+
```
nit:
```suggestion
```

joestringer commented on Documentation/network/pod-mac-address.rst:
```diff
@@ -0,0 +1,69 @@
+container it will not collide with any other MAC addresses assigned to other
+pods on the same node.
+
+
```
nit:
```suggestion
```

joestringer reviewed (changes requested):
Thanks for the writing up the user-facing docs on this, I think it makes sense. I've made a few wording suggestions to simplify and streamline the descriptions to be more in line with the docs style of the rest of the docs.

joestringer commented on Documentation/network/pod-mac-address.rst:
```diff
@@ -0,0 +1,59 @@
+
+Cilium will configure the MAC address for the primary interface inside a
+Pod if you specify the MAC address in the ``cni.cilium.io/mac-address`` before
+deploying the Pod. This MAC address is isolated to the container so it will
```
nit: 
```suggestion
Pod if you specify the MAC address in the ``cni.cilium.io/mac-address`` annotation before
deploying the Pod. This MAC address is isolated to the container so it will
```

joestringer commented on test/k8s/pod_mac_address.go:
```diff
@@ -0,0 +1,70 @@
+			Expect(demoPods.Items).To(HaveLen(1))
+			By("Checking the pod's mac address")
+			kubectl.ExecPodCmd(helpers.DefaultNamespace, demoPods.Items[0].Name,
+				fmt.Sprintf("ip link show |grep \"%s\" | wc -l", demoPods.Items[0].Annotations[annotation.PodAnnotationMAC])).ExpectSuccess("1")
```
Please replace the `"1"` parameter to `ExpectSuccess()` here with a meaningful error message that can be reported to developers if the test fails.

Useful context would be the `demoPods.Items[0].Name` so the developer could understand which pod was attempted to be exec'd into, as well as a brief description of what this command is trying to do.

Such a suggestion may look something like this: `"cannot find configured mac address %s in 'ip link show' output from pod %s"`.

chaunceyjiang replied:
Updated.

joestringer reviewed (changes requested):
Just a couple more minor nits, we're almost there. Thanks :pray:

aanm reviewed (approved):
Thank you for the PR.

@chaunceyjiang can you rebase to fix the conflicts?

aanm reviewed (approved):
The tests were successful on my private branch.
